### PR TITLE
Fix Imgui: account for FramebufferScale

### PR DESCRIPTION
**Pull request by zjanetzek**
This should make imgui scale correctly on Mac Retina displays (when used with glfw and GLFW_COCOA_RETINA_FRAMEBUFFER is set false).

### CODE REVIEW

**Comment by TheMostDiligent** [comment]
Thanks for the PR. Do samples and tutorials work OK with this change?

**Comment by zjanetzek** [comment]
A quick search gave no results for `ImGui::GetIO().DisplayFramebufferScale` in samples, but I'll give it a test. As DisplayFramebufferScale defaults to 1 all the multiplications in this PR should make no difference. The only usecase of DisplayFramebufferScale is to support this low-res mode on retina displays where everything is in css pixels or something :D

**Comment by TheMostDiligent** [comment]
@zjanetzek I do remember about this PR. I just don't have time to test it at the moment, but will get to it later.

**Comment by TheMostDiligent** [comment]
@zjanetzek Hi, I recently made some major changes to imgui to support orientation changes on Android. I was going to merge you changes, but I am not sure they will work now. I added few assertions and I believe they will fail. Can you get the latest version and see what's going on?
Also, on my Mac I don't see FramebufferScale being set to 2. FYI, on retina displays I set [global font scale to 2](https://github.com/DiligentGraphics/DiligentTools/blob/0aea746b8fc6ed720e5a08a23ba8a1ef6b103734/Imgui/src/ImGuiImplMacOS.mm#L41) instead of framebuffer scale.

**Comment by dolphineye** [comment]
Tutorials & Samples are OK on iOS and macOS with the fix suggested above. Let me know if you need any additional information.

**Comment by TheMostDiligent** [comment]
@dolphineye Do you call `NewFrame` like [here](https://github.com/DiligentGraphics/DiligentSamples/blob/master/SampleBase/src/SampleApp.cpp#L713)? My concern is that Imgui implementation on [Mac](https://github.com/DiligentGraphics/DiligentTools/blob/master/Imgui/src/ImGuiImplMacOS.mm#L56) and [iOS](https://github.com/DiligentGraphics/DiligentTools/blob/master/Imgui/src/ImGuiImplIOS.mm#L57) then sets DisplaySize to be the same as physical render surface size. Because of that I don't quite understand why only ClipRect needs to be scaled, not the viewport in particular. I understand that scaling by 1 is not going to change much, but it would be nice to understand what is going on with this framebuffer scale.

**Comment by dolphineye** [comment]
> @dolphineye Do you call `NewFrame` like [here](https://github.com/DiligentGraphics/DiligentSamples/blob/master/SampleBase/src/SampleApp.cpp#L713)? My concern is that Imgui implementation on [Mac](https://github.com/DiligentGraphics/DiligentTools/blob/master/Imgui/src/ImGuiImplMacOS.mm#L56) and [iOS](https://github.com/DiligentGraphics/DiligentTools/blob/master/Imgui/src/ImGuiImplIOS.mm#L57) then sets DisplaySize to be the same as physical render surface size. Because of that I don't quite understand why only ClipRect needs to be scaled, not the viewport in particular. I understand that scaling by 1 is not going to change much, but it would be nice to understand what is going on with this framebuffer scale.

Hold up, I used the identity transform instead of that of the swap chain. Let me try with it and get back to you. If this is not OK, I will try another way and get back to you after I have more context. Thanks for helping!

**Comment by dolphineye** [comment]
@TheMostDiligent Using the swap chain desc's `PreTransform` didn't solve the problem. I'm going to try another way around

**Comment by TheMostDiligent** [comment]
> Using the swap chain desc's `PreTransform` didn't solve the problem. I'm going to try another way around

`PreTransform` may be something other than identity only on Vulkan on Android. My main question was rather about the [DisplaySize being initialized the same as render surface size](https://github.com/DiligentGraphics/DiligentTools/blob/master/Imgui/src/ImGuiImplMacOS.mm#L56). Becasue of that I don't understand why `FramebuffScale` should only be applied to clip rects. Also, if imgui believes that all coordinates must be doubled, I don't understand how it would correctly handle input coordinates.

**Comment by dolphineye** [comment]
Hey, sorry for the long delay for the reply :)

So as far as I could check in the ImGUI examples, it seems that the `io.DisplaySize` must be the resolution size of the screen, and this size should be then multiplied by the `FrameBufferScale` when rendering draw data. For example:

```cpp
void    ImGui_ImplOpenGL3_RenderDrawData(ImDrawData* draw_data)
{
    // Avoid rendering when minimized, scale coordinates for retina displays (screen coordinates != framebuffer coordinates)
    int fb_width = (int)(draw_data->DisplaySize.x * draw_data->FramebufferScale.x);
    int fb_height = (int)(draw_data->DisplaySize.y * draw_data->FramebufferScale.y);
    if (fb_width <= 0 || fb_height <= 0)
        return;

    // ....

    ImVec2 clip_scale = draw_data->FramebufferScale;

    // ....

    for (int cmd_i = 0; cmd_i < cmd_list->CmdBuffer.Size; cmd_i++)
    {
        ImVec4 clip_rect;
        clip_rect.x = (pcmd->ClipRect.x - clip_off.x) * clip_scale.x;
        clip_rect.y = (pcmd->ClipRect.y - clip_off.y) * clip_scale.y;
        clip_rect.z = (pcmd->ClipRect.z - clip_off.x) * clip_scale.x;
        clip_rect.w = (pcmd->ClipRect.w - clip_off.y) * clip_scale.y;
    }

    // ....
}
```

I'm in the process of migrating to the ImGUI's `docking` branch, and using `DisplaySize` = `WindowSize * FrameBufferScale` gives inconsistent results.

In my case, the [fix suggested here](https://github.com/DiligentGraphics/DiligentTools/pull/8#issuecomment-621416685) sufficed to give me consistent results on all platforms (desktop and mobile) and works seamlessly on multi-monitor setups on desktop.

So, long story short, I think that the Diligent ImGUI's implementation is incorrectly using ImGUI `DisplaySize` and `FrameBufferSize` values. However, this should be quite easy and quick to fix. It would probably be necessary to also fix the different platform-specific samples but it should be pretty quick.

**Comment by dolphineye** [comment]
Just to complement my previous message, here's the diff that gives the expected results:

```diff

---
 Imgui/src/ImGuiImplDiligent.cpp | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/Imgui/src/ImGuiImplDiligent.cpp b/Imgui/src/ImGuiImplDiligent.cpp
index 1db7de5..efea59f 100644
--- a/Imgui/src/ImGuiImplDiligent.cpp
+++ b/Imgui/src/ImGuiImplDiligent.cpp
@@ -533,12 +533,12 @@ void ImGuiImplDiligent_Internal::RenderDrawData(IDeviceContext* pCtx, ImDrawData
         pCtx->SetBlendFactors(blend_factor);
 
         Viewport vp;
-        vp.Width    = static_cast<float>(m_RenderSurfaceWidth);
-        vp.Height   = static_cast<float>(m_RenderSurfaceHeight);
+        vp.Width    = static_cast<float>(m_RenderSurfaceWidth) * pDrawData->FramebufferScale.x;
+        vp.Height   = static_cast<float>(m_RenderSurfaceHeight) * pDrawData->FramebufferScale.y;
         vp.MinDepth = 0.0f;
         vp.MaxDepth = 1.0f;
         vp.TopLeftX = vp.TopLeftY = 0;
-        pCtx->SetViewports(1, &vp, m_RenderSurfaceWidth, m_RenderSurfaceHeight);
+        pCtx->SetViewports(1, &vp, m_RenderSurfaceWidth * pDrawData->FramebufferScale.x, m_RenderSurfaceHeight * pDrawData->FramebufferScale.y);
     };
 
     SetupRenderState();
@@ -568,10 +568,10 @@ void ImGuiImplDiligent_Internal::RenderDrawData(IDeviceContext* pCtx, ImDrawData
                 // Apply scissor/clipping rectangle
                 float4 ClipRect //
                     {
-                        pcmd->ClipRect.x - pDrawData->DisplayPos.x,
-                        pcmd->ClipRect.y - pDrawData->DisplayPos.y,
-                        pcmd->ClipRect.z - pDrawData->DisplayPos.x,
-                        pcmd->ClipRect.w - pDrawData->DisplayPos.y //
+                        (pcmd->ClipRect.x - pDrawData->DisplayPos.x) * pDrawData->FramebufferScale.x,
+                        (pcmd->ClipRect.y - pDrawData->DisplayPos.y) * pDrawData->FramebufferScale.y,
+                        (pcmd->ClipRect.z - pDrawData->DisplayPos.x) * pDrawData->FramebufferScale.x,
+                        (pcmd->ClipRect.w - pDrawData->DisplayPos.y) * pDrawData->FramebufferScale.y //
                     };
                 // Apply pretransform
                 ClipRect = TransformClipRect(pDrawData->DisplaySize, ClipRect);
@@ -583,7 +583,7 @@ void ImGuiImplDiligent_Internal::RenderDrawData(IDeviceContext* pCtx, ImDrawData
                         static_cast<Int32>(ClipRect.z),
                         static_cast<Int32>(ClipRect.w) //
                     };
-                pCtx->SetScissorRects(1, &r, m_RenderSurfaceWidth, m_RenderSurfaceHeight);
+                pCtx->SetScissorRects(1, &r, m_RenderSurfaceWidth * pDrawData->FramebufferScale.x, m_RenderSurfaceHeight * pDrawData->FramebufferScale.y);
 
                 // Bind texture, Draw
                 auto* texture_srv = reinterpret_cast<ITextureView*>(pcmd->TextureId);
-- 
2.19.0
```

I have not checked on Android if this is working correctly with the surface pre-transform. I guess that if there were to be problem, it could easily be fixed similarly to the diff from above.

Also, such fix implies that calls to `ImGuiImplDiligent::NewFrame` (and its subclasses) do not pre-scale the render surface's size by the DPI scale. I would understand that you would not like to modify your API about that, but as far as I could test the docking branch, using a render surface size pre-scaled by the DPI scale enforced me to "patch" the `DrawData` before rendering it with `ImGuiImplDiligent::RenderDrawData`.

Let me know if you need any additional info, Id be happy to help!

**Comment by TheMostDiligent** [comment]
Closing as this is superseded by https://github.com/DiligentGraphics/DiligentTools/pull/10
